### PR TITLE
Feature devise and authorization

### PR DESCRIPTION
**Pull request by IsmailChangezi**
## Description
- Add Devise
- Add Authorization

### CODE REVIEW

**Comment by KDlamini** [comment]
## Status: Invalid 🔴 

### Hi @IsmailChangezi  🙋‍♀️,  

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request
 
You have submitted a project, that does not have equal commits from your coding partner.

- Please note that this milestone is a [pair-programming activity](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/blog_app_devise.md#work-and-submission-mode). Kindly ensure genuine collaboration between partners, the commit history should reflect that everyone has contributed equally to the project.
- You might need to start a new branch and pull a request to ensure that your partner has the opportunity to make significant contributions to this project. We will not accept a facade collaboration where the coding partner only makes one commit that does not add any significant feature.

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏
 
Cheers 🥂  and Happy coding!!! 👯 
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @KDlamini  in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_Invalid Code Review Request **does not count** into the code reviews limit._

**Review by Qoosim — COMMENTED** [review]
#  *Status: Invalid Request* :red_circle: 

Hi @IsmailChangezi, 

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.



## Invalid Code Review Request

You have submitted a project, that does not have any contribution from your partner. Kindly ask your partner to make at least one major commit not, modify readme kind of commit or attach an approval from student success stated you to work on this project solo.

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

**Comment by IsmailChangezi** [comment]
## Important
I am working alone this week so the requirement of pair programming is not applicable on me. 
##### Please keep this in mind while reviewing this PR

**Comment by Qoosim** [comment]
> ## Important
> I am working alone this week so the requirement of pair programming is not applicable on me.
> 
> ##### Please keep this in mind while reviewing this PR

@IsmailChangezi, kindly reply to my message on slack.
I simply asked you to attach the mail received from Student Success stated that you can work solo.
I reclaimed your PR by the way and will hold on for say 10 mins for you to respond to this message whether here or on slack.
The response determines whether I will invalidate the PR again or proceed with the review. Thanks

**Review by Qoosim — COMMENTED** [review]
#  *Status: Invalid Request* :red_circle: 

Hi @IsmailChangezi, 

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.



## Invalid Code Review Request

You have submitted a project, that is supposed to commits from both partners but have commits only from one partner.

**NOTE:** Kindly attach the mail you received from the `Student Success` ordered you to work solo for the week.

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

**Comment by IsmailChangezi** [comment]
# This is the Screenshot of the email sent by the Microverse

File: app/models/user.rb [hunk 1]
```diff
@@ -1,12 +1,16 @@
...
-    has_many :posts, foreign_key: 'author_id'
-    has_many :comments, foreign_key: 'author_id'
-    has_many :likes, foreign_key: 'author_id'
-  
-    validates :name, presence: true
-    validates :posts_counter, numericality: { greater_than_or_equal_to: 0 }
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
```
Comment (ShadyShawkat) [inline app/models/user.rb]: - [ ] Kindly add `:confirmable` to the user model and make sure the user can confirm their accounts before logging in, as per [this](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/blog_app_devise.md#:~:text=Ask%20for%20confirmation%20of%20email.) requirement.

File: app/models/ability.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+  def initialize(user)
+    # Define abilities for the user here. For example:
+    #
+    return unless user.present?
+
+     can :read, :all
+     can :create, [Post, Comment]
+
+     if user.role = 'admin'
```
Comment (Deepakdanger) [inline app/models/ability.rb:14]: - [ ] [ OPTIONAL] You can also create a method in user controller named`admin?` and use it here as  `user.admin?`, This way you can use this method anywhere.
```ruby
def admin?
     role == 'admin'
end
```

**Review by Deepakdanger — APPROVED** [review]
Hi @IsmailChangezi , 

### Status: Approved ✔️

Your project is complete!  There is nothing else to say other than... it's time to merge it  🚢 
Congratulations! 🎉 

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- x

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/devise/sessions/new.html.erb [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+      <%= f.label :remember_me %>
+    </div>
+  <% end %>
+
+  <div class="actions">
+    <%= f.submit "Log in" %>
+  </div>
+<% end %>
+
+<%= render "devise/shared/links" %>
```
Comment (ShadyShawkat) [inline app/views/devise/sessions/new.html.erb:26]: - [ ] Kindly make sure that the design of the login and sign-up forms matches the [given wireframe](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/sneak_peek.md), as per [this](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/blog_app_devise.md#:~:text=Modify%20the%20views%20in%20devise%20for%20registration%20and%20login%20to%20match%20the%20wireframes%20from%20the%20Sneak%20Peek%20and%20your%20styling.) requirement.

image

**Review by ShadyShawkat — CHANGES_REQUESTED** [review]
## Hi @IsmailChangezi, 

Good job so far! 👍🏻 
There are some issues that you still need to work on to go to the next project but you are almost there!

## To Highlight
- Correct use of GitFlow ✔️ 
- Descriptive commit messages ✔️ 
- Professional looking pull request ✔️ 

Image

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by tugsem** [comment]
## Invalid Code Review Request ❌ 

Hi @IsmailChangezi , 

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

<h2>You have submitted a project, that does not have the required changes. This is an old pull request for the prior milestone. You should open a new PR with the changes(Add API endpoints)</h2>

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

**Review by Alfredbis29 — APPROVED** [review]
# STATUS: APPROVED   
 ## Hi  @IsmailChangezi 
Your project is complete! There is nothing else to say other than... Great job clap Congratulations! :tada:
 
 TO HIGHLIGHT:
 
 * Your project is professional :white_check_mark:
 * Good Job :+1:
 
 Now you can merge this branch and move on 1st_place_medal: HAPPY CODING AND GOOD LUCK WITH THE COMING ONES :crossed_fingers: :


 
 ## Optional suggestions

- [x] N/A

 _Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._
 
Cheers and Happy coding! :clap:  :clap: :clap:

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me @Alfredbis29 in your question so I can receive the notification.
 
 _As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**